File: gob/optimizers/cpp_optimizers/src/optimizers/particles/common-noise/CN_SBS.cc
/*
 * Created in 2025 by Gaëtan Serré
 */

#include "optimizers/particles/common-noise/common_noise.hh"
#include "optimizers/particles/common-noise/CN_SBS.hh"

void CN_SBS::set_stop_criterion(double stop_criterion)
{
  this->stop_criterion = stop_criterion;
  this->has_stop_criterion = true;
  this->base_opt.set_stop_criterion(stop_criterion);
}

result_eigen CN_SBS::minimize(function<double(dyn_vector)> f)
{
  Common_Noise cn(&this->base_opt, this->gamma, this->lambda, this->delta, static_cast<NoiseType>(this->moment), "CN_SBS");
  return cn.minimize(f);
}

File: gob/optimizers/particles/CN_SBS.py
#
# Created in 2024 by Gaëtan Serré
#

from ..optimizer import Optimizer
from ..cpp_optimizers import CN_SBS as CCN_SBS


class CN_SBS(Optimizer):
    """
    Interface for the Common noise SBS optimizer.

    Parameters
    ----------
    bounds : ndarray
        The bounds of the search space.
    n_particles : int
        The number of particles.
    iter : int
        The number of iterations.
    dt : float
        The time step.
    sigma : float
        The kernel bandwidth.
    gamma : float
        The coefficient for the common noise.
    ``lambda_`` : float
        The regularization parameter for the common noise.
    delta : float
        The parameter for the Bessel process.
    moment : str
        The type of moment used for the common noise ("M1" | "M2" | "VAR").
    verbose : bool
        Whether to print information about the optimization process.
    """

    def __init__(
        self,
        bounds,
        n_particles=200,
        iter=100,
        dt=10,
        sigma=0.1,
        gamma=1,
        lambda_=0,
        delta=2.1,
        moment="M2",
        verbose=False,
    ):
        super().__init__("CN_SBS", bounds)

        match moment:
            case "M1":
                moment = 0
            case "M2":
                moment = 1
            case "VAR":
                moment = 2
            case _:
                raise ValueError(
                    'Invalid moment type. Choose from "M1", "M2", or "VAR".'
                )

        self.c_opt = CCN_SBS(
            bounds, n_particles, iter, dt, sigma, gamma, lambda_, delta, moment
        )
        self.verbose = verbose

    def minimize(self, f):
        if self.verbose:
            f = self.verbose_function(f)
        return self.c_opt.minimize(f)

    def set_stop_criterion(self, stop_criterion):
        self.c_opt.set_stop_criterion(stop_criterion)

    def __del__(self):
        del self.c_opt
